### ISSUE TITLE
support breaking changes of OpenAPI 3.1.0

### ISSUE THREAD

**Issue by ysndr** [oxidecomputer/progenitor#762]
OpenAPI version 3.1.0 came with some breaking changes, notably, the introduction of a `null` type and removal of the `nullable` attribute.

We can work around this by patching these parts of a 3.1.0 spec to a 3.0.* spec, but i'm concerned about the implication that has on the validity of the document.

Is an extension of progenitor to support OpenAPI 3.1.0 being considered or worked on?

**Comment by ahl** [comment]
> Is an extension of progenitor to support OpenAPI 3.1.0 being considered or worked on?

Definitely considered. It's been low priority for us because of the generally slow uptake of 3.1. 3.1 also turns out to contain some extremely complicated features to support in particular with the adoption of JSON Schema 2020-12 and the `jsonSchemaDialect ` property that allows random other schemas to be used instead. In addition the `openapiv3` crate doesn't support 3.1--in part because the `schemars` crate doesn't support 2020-12.

In short: there's an expedient, but imprecise path forward that I've been deferring in the hope of doing the more complete things. Happy to discuss further if that would be helpful.

**Comment by ahl** [comment]
And I forgot: one important change in 2020-12 is how tuples are represented. Previously it was an array with `items` and `additionalItems`. In 2020-12 it became `prefixItems` and `items` where `items` is now always a singleton. It's a change for the better, but is also--obviously--incompatible.

**Comment by jayvdb** [comment]
Note openapi generator utoipa has switched to OpenAPI v3.1 ; c.f. https://github.com/juhaku/utoipa/#features

https://github.com/paperclip-rs/paperclip & https://github.com/GREsau/okapi are still 3.0.x

**Comment by metasim** [comment]
Yep, I’m here because of utoipa… :/

**Comment by lostb1t** [comment]
Another vote for 3.1 support. I use multiple api’s that are in 3.1

**Comment by FliegendeWurst** [comment]
I've used https://github.com/apiture/openapi-down-convert to convert 3.1.0 -> 3.0.3 with success. It doesn't handle all new constructs, but works for most specs.

**Comment by danwt** [comment]
Voting

**Comment by dzmitry-lahoda** [comment]
will donate 43USD equivalent in crypto for 3.1 support (i like `const` feature)

**Comment by ClemaX** [comment]
If someone is willing to accept I will be bidding 50USD in ETH on this issue too.

**Comment by Eveeifyeve** [comment]
Well as a CEO on behalf of [TeaClient (Our quality of Life Minecraft Client)](https://teaclient) we are needing this support too for [our registry CLI](https://github.com/TeaClientMC/Registry)! I will be getting work on this support to better help opensource, but also better help our registry.

We considered using JSON schemas, but however it wasn't convenient than one schema than Open API. We don't have a crypto wallet to accept the bids/donates, we can get back to you via [our discord](https://teaclient.com/discord) dming me or @TheTCraft to find a official way of accepting these bids/donations.

**Comment by Eveeifyeve** [comment]
Looking at this it looks like this issue needs to be solved first https://github.com/oxidecomputer/typify/issues/886

**Comment by TheCactusBlue** [comment]
I will take it on, if no one else is on this
